internal/provider/users_data_source.go

```diff
@@ -0,0 +1,175 @@
+	"type":       types.StringType,
+	"email":      types.StringType,
+	"created_at": types.StringType,
+}
+
+func (d *UsersDataSource) Metadata(ctx context.Context, req datasource.MetadataRequest, resp *datasource.MetadataResponse) {
+	resp.TypeName = req.ProviderTypeName + "_users"
+}
+
+func (d *UsersDataSource) Schema(ctx context.Context, req datasource.SchemaRequest, resp *datasource.SchemaResponse) {
+	resp.Schema = schema.Schema{
+		MarkdownDescription: "Humanitec users",
+
+		Attributes: map[string]schema.Attribute{
+			"id": schema.StringAttribute{
+				Computed: true,
+			},
+			"filter": schema.SingleNestedAttribute{
+				Attributes: map[string]schema.Attribute{
+					"id": schema.StringAttribute{
+						Optional: true,
+					},
+					"name": schema.StringAttribute{
+						Optional: true,
+					},
+					"email": schema.StringAttribute{
+						Optional: true,
+					},
+				},
+				Optional: true,
+			},
+			"users": schema.ListAttribute{
+				ElementType: types.ObjectType{
+					AttrTypes: userAttrTypes,
+				},
+				Computed: true,
+			},
+		},
+	}
+}
+
+func (d *UsersDataSource) Configure(ctx context.Context, req datasource.ConfigureRequest, resp *datasource.ConfigureResponse) {
+	// Prevent panic if the provider has not been configured.
+	if req.ProviderData == nil {
+		return
+	}
+
+	resdata, ok := req.ProviderData.(*HumanitecData)
+
+	if !ok {
+		resp.Diagnostics.AddError(
+			"Unexpected Resource Configure Type",
+			fmt.Sprintf("Expected *http.Client, got: %T. Please report this issue to the provider developers.", req.ProviderData),
+		)
+
+		return
+	}
+
+	d.client = resdata.Client
+	d.orgId = resdata.OrgID
+}
+
+func (d *UsersDataSource) Read(ctx context.Context, req datasource.ReadRequest, resp *datasource.ReadResponse) {
+	var data UsersDataSourceModel
+
+	// Read Terraform configuration data into the model
+	resp.Diagnostics.Append(req.Config.Get(ctx, &data)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	httpResp, err := d.client.ListUserRolesInOrgWithResponse(ctx, d.orgId)
+	if err != nil {
+		resp.Diagnostics.AddError(HUM_CLIENT_ERR, fmt.Sprintf("Unable to list users, got error: %s", err))
+		return
+	}
+	if httpResp.StatusCode() != 200 {
+		resp.Diagnostics.AddError(HUM_API_ERR, fmt.Sprintf("Unable to list users, unexpected status code: %d, body: %s", httpResp.StatusCode(), httpResp.Body))
+		return
+	}
+
+	var id, name, email *string
+	if !data.Filter.IsNull() {
+		var filter UsersFilterDataSourceModel
+		resp.Diagnostics.Append(data.Filter.As(ctx, &filter, basetypes.ObjectAsOptions{})...)
+		if resp.Diagnostics.HasError() {
+			return
+		}
+
+		id = filter.Id.ValueStringPointer()
+		name = filter.Name.ValueStringPointer()
+		email = filter.Email.ValueStringPointer()
+	}
+	userIds := []string{}
+	users := []basetypes.ObjectValue{}
+	for _, userRole := range *httpResp.JSON200 {
+		if (id == nil || userRole.Id == *id) && (name == nil || userRole.Name == *name) && (email == nil || *userRole.Email == *email) {
```

Comment:
I found this a bit hard to reason about, maybe it would make sense to extract this into a function like `matchesFilter(filter *UsersFilterDataSourceModel, userRole) bool`?

Reply:
Refactored, hope it's more clear now